store/idp.go

```diff
@@ -0,0 +1,268 @@
+package store
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/bytebase/bytebase/api"
+	"github.com/bytebase/bytebase/common"
+)
+
+// IdentityProviderMessage is the mssage for identity provider.
+type IdentityProviderMessage struct {
+	ResourceID string
+	Title      string
+	Domain     string
+	Type       api.IdentityProviderType
+	Config     string
+	// The following fields are output only and not used for creating.
+	UID     int
+	Deleted bool
+}
+
+// FindIdentityProviderMessage is the message for finding identity providers.
+type FindIdentityProviderMessage struct {
+	// We should only set either UID or ResourceID.
+	// Deprecate UID later once we fully migrate to ResourceID.
+	UID         *int
+	ResourceID  *string
+	ShowDeleted bool
+}
+
+// UpdateIdentityProviderMessage is the message for updating an identity provider.
+type UpdateIdentityProviderMessage struct {
+	UpdaterID  int
+	ResourceID string
+
+	Title  *string
+	Domain *string
+	Config *string
+	Delete *bool
+}
+
+// CreateIdentityProvider creates an identity provider.
+func (s *Store) CreateIdentityProvider(ctx context.Context, create *IdentityProviderMessage, creatorID int) (*IdentityProviderMessage, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, FormatError(err)
+	}
+	defer tx.Rollback()
+
+	identityProvider := &IdentityProviderMessage{
+		ResourceID: create.ResourceID,
+		Title:      create.Title,
+		Domain:     create.Domain,
+		Type:       create.Type,
+		Config:     create.Config,
+	}
+	if err := tx.QueryRowContext(ctx, `
+			INSERT INTO idp (
+				creator_id,
+				updater_id,
+				resource_id,
+				name,
+				domain,
+				type,
+				config
+			)
+			VALUES ($1, $2, $3, $4, $5, $6, $7)
+			RETURNING id
+		`,
+		creatorID,
+		creatorID,
+		create.ResourceID,
+		create.Title,
+		create.Domain,
+		create.Type,
+		create.Config,
+	).Scan(
+		&identityProvider.UID,
+	); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, common.FormatDBErrorEmptyRowWithQuery("failed to create identity provider")
+		}
+		return nil, FormatError(err)
+	}
+
+	if err := tx.Commit(); err != nil {
+		return nil, FormatError(err)
+	}
+
+	return identityProvider, nil
+}
+
+// GetIdentityProvider gets an identity provider.
+func (s *Store) GetIdentityProvider(ctx context.Context, find *FindIdentityProviderMessage) (*IdentityProviderMessage, error) {
+	// We will always return the resource regardless of its deleted state.
```

Comment:
Shouldn't the caller make the decision whether to include the deleted entry?

Reply:
https://google.aip.dev/164#list-and-get

Reply:
I see, we should add link in code comments because it gives no context why we choose to do so.